Add new `Rails/Env` cop

cdudas17 opened the pull request:
**Replace this text with a summary of the changes in your PR.
The more detailed you are, the better.**

-----------------

Before submitting the PR make sure the following are checked:

* [ ] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [ ] Wrote [good commit messages][1].
* [ ] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [ ] Feature branch is up-to-date with `master` (if not - rebase it).
* [ ] Squashed related commits together.
* [ ] Added tests.
* [ ] Ran `bundle exec rake default`. It executes all tests and runs RuboCop on its own code.
* [ ] Added an entry (file) to the [changelog folder](https://github.com/rubocop/rubocop-rails/blob/master/changelog/) named `{change_type}_{change_description}.md` if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format) for details.
* [ ] If this is a new cop, consider making a corresponding update to the [Rails Style Guide](https://github.com/rubocop/rails-style-guide).

[1]: https://chris.beams.io/posts/git-commit/

Earlopain commented on lib/rubocop/cop/rails/env.rb:
```diff
@@ -0,0 +1,31 @@
+      # @example
+      #
+      #   # bad
+      #   Rails.env.development? || Rails.env.production? || Rails.env.local?
```
`local` already encompasses `development`. Can you write this differently?

cdudas17 replied:
How does `Rails.env.development? || Rails.env.production? || Rails.env.test?` look? Or did you have something particular in mind?

Earlopain replied:
Sure, that seems fine. You could also drop `development` instead, I don't have a preference.

Earlopain reviewed (commented):
Thanks for the PR. I agree with the general idea of what the cop is about but question if it should be enabled by default.

Checking the env directly seems like a somewhat common thing to do, and resolving the offense is not trivial. I can also see how some would just not care about feature-flags or similar at all, especially smaller applications. You did write in your issue that this is explicitly about large apps. In my opinion the cop should be opt-in instead.

----

`Rails.env.capitalize` is a offense, can be used to display in a dashboard of sorts. I think it probably fine to only display warnings for predicate methods.

Earlopain commented on CHANGELOG.md:
```diff
@@ -1152,3 +1152,4 @@
 [@fwolfst]: https://github.com/fwolfst
 [@maxprokopiev]: https://github.com/maxprokopiev
 [@ytjmt]: https://github.com/ytjmt
+[@cdudas17]: https://github.com/cdudas17
```
This file is autogenerated, please remove. When a release is made it is properly updated.

Earlopain commented on changelog/new_env_cop.md:
```diff
@@ -0,0 +1 @@
+* [#1375](https://github.com/rubocop/rubocop-rails/pull/1375): Add new `Rails/Env` cop. ([@cdudas17][])
```
Can you instead link the issue you openend?

cdudas17 commented:
@Earlopain That makes sense that the cop should be opt-in. I changed the config default from `pending` to `false`. I also updated some other comments you had. `Rails.env.capitalize` is no longer an offense, and I added a test for it. We are looking for predicate methods now and including those methods as the offending node.

Earlopain commented:
Hm, yeah. That's a good point there. I don't particularly like the solution, `Rails/UnknownEnv` already knows about the available envs (courtesy of the user configuring it) and it feels like that should have an affect here.

But I don't believe there is precedence for overarching configuration like that. So, I'd just leave it like you have it now. Though I think your list includes a cusom core extension? `contains_mb4_chars` doesn't sound familiar and I find no reference to it anywhere. `colorized?` maybe from a gem called `colorized`? Here are my results:
```rb
require "active_support/all"
("".inquiry.methods - Object.instance_methods).select { |m| m.to_s.end_with?('?') }
=> [:unicode_normalized?, :starts_with?, :exclude?, :ends_with?, :empty?, :acts_like_string?, :is_utf8?, :casecmp?, :match?, :start_with?, :end_with?, :valid_encoding?, :ascii_only?, :between?]
```

Methods like that seem fine to not appear, as long as the common ones like `start_with?`/`end_with?` are there. Notice how I use `instance_methods` instead, which makes `include?` appear as a potential canidate. It previously got filtered out because `Object` through `Module` responds to it.

cdudas17 commented:
@Earlopain The allow list isn't my favorite solution either but I'm not sure what a better solution would be. I'll defer to your decision about overreaching in configuration. I updated the allow list to be yours and confirmed I had the same list of predicate methods in a more vanilla rails app.

Earlopain commented:
This looks good to me, I have no more comments. I'm not a maintainer, so in the end the harder decisions are (luckily) not up to me (:

cdudas17 commented:
@Earlopain is there workflow I'm missing to get this properly approved? I read https://github.com/rubocop/rubocop-rails/blob/master/CONTRIBUTING.md and think I'm following all the steps. Do I need to request from someone specifically or am I missing a step that you're aware of? You're my only point of contact in this repo so far 😅

Earlopain commented:
Sorry for the late response, I forgot. You didn't miss anything, it's just that adding new cops here can be a bit of a slog. If you look at the currently open PRs, most are for new cops. Can't give you anything other than being patient I'm afraid.
